src/algo/traversal/dfs/mod.rs

```diff
@@ -29,57 +27,72 @@ pub enum EdgeType {
     CrossEdge,
 }
 
-pub enum Event<'a, G: NodeIdMapProvider> {
-    Begin(&'a State<G>, NodeId),
-    Discover(&'a State<G>, NodeId),
-    Finish(&'a State<G>, NodeId),
-    End(&'a State<G>),
-    VisitEdge(&'a State<G>, EdgeType, NodeId, NodeId),
+pub enum Event<'a, I: IdMap> {
+    Begin(&'a State<I>, NodeId),
+    Discover(&'a State<I>, NodeId),
+    Finish(&'a State<I>, NodeId),
+    End(&'a State<I>),
+    VisitEdge(&'a State<I>, EdgeType, NodeId, NodeId),
 }
 
-pub struct State<G: NodeIdMapProvider> {
-    pub id_map: <G as NodeIdMapProvider>::NodeIdMap,
+pub struct State<I: IdMap> {
+    pub id_map: I,
 
     pub time: usize,
     pub discover: Vec<usize>,
     pub finished: Vec<usize>,
     pub parent: Vec<usize>,
-
-    pub(crate) phantom_g: PhantomData<G>,
 }
 
-pub struct DFS<'a, G>
+pub struct DFS<'a, G, I = DefaultIdMap>
 where
-    G: NodeProvider + NodeIdMapProvider,
+    G: NodeProvider,
+    I: IdMap,
 {
     graph: &'a G,
     starting_nodes: Vec<NodeId>,
-    state: State<G>,
+    state: State<I>,
 }
 
 impl<'a, G> DFS<'a, G>
 where
-    G: NodeProvider + NodeIdMapProvider,
+    G: NodeProvider,
 {
     pub fn init(graph: &'a G) -> Self {
-        Self::with_starters(graph, None.into_iter())
+        Self::with_ids_and_starters(graph, DefaultIdMap::new(graph), None.into_iter())
     }
 
     pub fn with_starters(graph: &'a G, starting_nodes: impl Iterator<Item = NodeId>) -> Self {
+        Self::with_ids_and_starters(graph, DefaultIdMap::new(graph), starting_nodes)
+    }
+}
+
+impl<'a, G, I> DFS<'a, G, I>
+where
+    G: NodeProvider,
+    I: IdMap,
+{
+    pub fn with_ids(graph: &'a G, id_map: I) -> Self {
+        Self::with_ids_and_starters(graph, id_map, None.into_iter())
+    }
```

Comment:
It may be that the provided `id_map` is not for the provided `graph`. Why don't we build the id map using `I::new(graph)` to be sure?
Is it for giving the user the option to reuse its previously created id map(for performance)?

Reply:
Here I try to avoid the need to specify type as `DFS::<_, AnotherIdMap>::with_ids(&graph)` because the type cannot be inferred.
It follows the convention like [HashMap::with_capacity_and_hasher](https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.with_capacity_and_hasher).